searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp

```diff
@@ -472,6 +480,89 @@ HnswIndex::trim_hold_lists(generation_t first_used_gen)
     _graph.links.trimHoldLists(first_used_gen);
 }
 
+void
+HnswIndex::compact_level_arrays(bool compact_memory, bool compact_address_space)
+{
+    auto context = _graph.nodes.compactWorst(compact_memory, compact_address_space);
+    uint32_t doc_id_limit = _graph.node_refs.size();
+    vespalib::ArrayRef<AtomicEntryRef> refs(&_graph.node_refs[0], doc_id_limit);
+    context->compact(refs);
+}
+
+void
+HnswIndex::compact_link_arrays(bool compact_memory, bool compact_address_space)
+{
+    auto context = _graph.links.compactWorst(compact_memory, compact_address_space);
+    uint32_t doc_id_limit = _graph.node_refs.size();
+    for (uint32_t doc_id = 1; doc_id < doc_id_limit; ++doc_id) {
+        EntryRef level_ref = _graph.node_refs[doc_id].load_relaxed();
+        if (level_ref.valid()) {
+            vespalib::ArrayRef<AtomicEntryRef> refs(_graph.nodes.get_writable(level_ref));
+            context->compact(refs);
+        }
+    }
+}
+
+bool
+HnswIndex::consider_compact_level_arrays(const CompactionStrategy& compaction_strategy)
+{
+    size_t used_bytes = _cached_level_arrays_memory_usage.usedBytes();
+    size_t dead_bytes = _cached_level_arrays_memory_usage.deadBytes();
+    bool compact_memory = compaction_strategy.should_compact_memory(used_bytes, dead_bytes);
+    size_t used_address_space = _cached_level_arrays_address_space_usage.used();
+    size_t dead_address_space = _cached_level_arrays_address_space_usage.dead();
+    bool compact_address_space = compaction_strategy.should_compact_address_space(used_address_space, dead_address_space);
+    if (compact_memory || compact_address_space) {
+        compact_level_arrays(compact_memory, compact_address_space);
+        return true;
+    }
+    return false;
+}
+
+bool
+HnswIndex::consider_compact_link_arrays(const CompactionStrategy& compaction_strategy)
+{
+    size_t used_bytes = _cached_link_arrays_memory_usage.usedBytes();
+    size_t dead_bytes = _cached_link_arrays_memory_usage.deadBytes();
+    bool compact_memory = compaction_strategy.should_compact_memory(used_bytes, dead_bytes);
+    size_t used_address_space = _cached_level_arrays_address_space_usage.used();
```

Comment:
Should this be _cached_link_arrays_address_space_usage instead?

Reply:
Fixed.